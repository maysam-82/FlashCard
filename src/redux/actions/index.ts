export * from './actionTypes';
export * from './stack';
