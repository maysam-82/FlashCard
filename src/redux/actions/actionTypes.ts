import { ISetStack, ILoadStacks, IAddStack } from './stack';

export enum ActionTypes {
	setStack,
	loadStacks,
	addStack,
}

export type StackAction = ISetStack;
export type StacksAction = ILoadStacks | IAddStack;
