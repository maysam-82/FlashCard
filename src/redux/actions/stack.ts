import { ActionTypes } from './actionTypes';

interface ICard {
	id: number;
	prompt: string;
	answer: string;
}

export interface IStack {
	title: string | null;
	id: number | null;
	cards: ICard[] | null;
}

export interface ISetStack {
	type: ActionTypes.setStack;
	payload: IStack;
}
export interface ILoadStacks {
	type: ActionTypes.loadStacks;
	payload: IStack[];
}

export const setStack = (stack: IStack): ISetStack => {
	return {
		type: ActionTypes.setStack,
		payload: stack,
	};
};
export const loadStacks = (stacks: IStack[]): ILoadStacks => {
	return {
		type: ActionTypes.loadStacks,
		payload: stacks,
	};
};
