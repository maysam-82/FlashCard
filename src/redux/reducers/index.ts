import { combineReducers } from 'redux';
import stackReducer from './stackReducer';
import { IStack } from '../actions';

export interface IStoreState {
	stack: IStack;
}

export const reducers = combineReducers<IStoreState>({
	stack: stackReducer,
});
