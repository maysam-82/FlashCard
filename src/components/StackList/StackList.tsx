import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { setStack, loadStacks, IStack } from '../../redux/actions/stack';
import stacks from '../../data/stacks.json';
import { IStoreState } from '../../redux/reducers/index';

export interface IProps {
	setStack: typeof setStack;
	loadStacks: typeof loadStacks;
	stacks: IStack[];
}
// using `export` to test disconnected component
export class StackList extends React.Component<IProps, {}> {
	componentDidMount() {
		if (this.props.stacks.length === 0) {
			this.props.loadStacks(stacks);
		}
	}

	renderStackList = () => {
		const { stacks, setStack } = this.props;
		if (this.props.stacks.length > 0) {
			return stacks.map((stack) => (
				<Link to="/stack" key={stack.id} onClick={() => setStack(stack)}>
					<h4>{stack.title}</h4>
				</Link>
			));
		}
	};
	render() {
		return <div>{this.renderStackList()}</div>;
	}
}

const mapStateToProps = (state: IStoreState) => ({
	stacks: state.stacks,
});

export default connect(mapStateToProps, { setStack, loadStacks })(StackList);
