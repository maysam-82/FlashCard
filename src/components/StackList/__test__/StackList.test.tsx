import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { StackList, IProps } from '../StackList';
import { IStack } from '../../../redux/actions';

const stack: IStack = {
	id: 500,
	title: 'test title stack',
	cards: [{ id: 2, prompt: 'test prompt', answer: 'test answer' }],
};

const mockSetStack = jest.fn();
const mockLoadStacks = jest.fn();
const props: IProps = {
	setStack: mockSetStack,
	loadStacks: mockLoadStacks,
	stacks: [stack],
};

describe('StackList component', () => {
	let component: ShallowWrapper<IProps, {}, StackList>;
	beforeEach(() => {
		component = shallow<StackList, IProps, {}>(<StackList {...props} />);
	});
	it('should render StackList properly', () => {
		expect(component).toMatchSnapshot();
	});
	// ComponentDidMount with not empty stacks
	describe('When StackList is mounted with not empty stacks', () => {
		it('should not call `loadStacks`', () => {
			expect(component.instance().props.loadStacks).toHaveBeenCalledTimes(0);
		});
		describe('When `Link` component is clicked', () => {
			beforeEach(() => {
				component.find('Link').simulate('click');
			});
			it('should call `setStack` function with `stack[0]` argument', () => {
				expect(component.instance().props.setStack).toHaveBeenCalledWith(
					props.stacks[0]
				);
			});
		});
	});
	// ComponentDidMount with an empty stacks
	describe('When StackList is mounted with an empty stacks', () => {
		beforeEach(() => {
			component = shallow<StackList, IProps, {}>(
				<StackList
					stacks={[]}
					setStack={props.setStack}
					loadStacks={props.loadStacks}
				/>
			);
		});
		it('should call `loadStacks`', () => {
			expect(component.instance().props.loadStacks).toHaveBeenCalledTimes(1);
		});
	});
});
