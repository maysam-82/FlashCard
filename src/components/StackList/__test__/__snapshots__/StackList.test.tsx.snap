// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackList component should render StackList properly 1`] = `
<div>
  <Link
    key="500"
    onClick={[Function]}
    to="/stack"
  >
    <h4>
      test title stack
    </h4>
  </Link>
</div>
`;
